### PR TITLE
[AMPGPU] Emit s_singleuse_vdst instructions when a register is used multiple times in the same instruction.

### PR DESCRIPTION
**Pull request by ScottEgerton**
Previously, multiple uses of a register within the same instruction were
being counted as multiple uses. This has been corrected to
only count as a single use as per the specification allowing for
more optimisation candidates.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 1]
```diff
@@ -83,26 +84,13 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
...
-          const auto Reg = Operand.getReg();
-
-          // Count the number of times each register is read.
-          if (Operand.readsReg())
-            for (const MCRegUnit &Unit : TRI->regunits(Reg))
-              RegisterUseCount[Unit]++;
+        // Gather a list of Registers used before updating use counts to avoid
+        // double counting registers that appear multiple times in a single
+        // MachineInstr.
+        DenseSet<MCRegUnit> RegistersUsed;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:90]: A DenseSet feels like a heavy way to track this

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 2]
```diff
@@ -114,6 +102,22 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
           for (const MCRegUnit &Unit : RegUnits)
             RegisterUseCount.erase(Unit);
         }
+
+        for (const auto &Operand : MI.all_uses()) {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:106]: What happens with regmask operands? What happens with call uses?

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 3]
```diff
@@ -83,26 +84,13 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
...
-        for (const auto &Operand : MI.operands()) {
-          if (!Operand.isReg())
-            continue;
-          const auto Reg = Operand.getReg();
-
-          // Count the number of times each register is read.
-          if (Operand.readsReg())
-            for (const MCRegUnit &Unit : TRI->regunits(Reg))
-              RegisterUseCount[Unit]++;
+        // Gather a list of Registers used before updating use counts to avoid
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:87]: Anything tracking per-instruction liveness should be split into a helper function. I feel like there's already something to do something similar for you somewhere, but I can't seem to find it. Is this just building a map from regunit to a 0-1-or-2 counter?

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 4]
```diff
@@ -114,6 +102,22 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
           for (const MCRegUnit &Unit : RegUnits)
             RegisterUseCount.erase(Unit);
         }
+
+        for (const auto &Operand : MI.all_uses()) {
+          const auto Reg = Operand.getReg();
+
+          // Count the number of times each register is read.
+          for (const MCRegUnit &Unit : TRI->regunits(Reg))
```
[the remark this answers is not in the document]
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:110]: I didn't realise this, I'll change this.

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 3 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:87]: Thanks for the feedback.
> Is this just building a map from regunit to a 0-1-or-2 counter?
Escentially, yes. We are counting the uses but the only values we care about are 0, 1 or >1.

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 2 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:106]: From the point of view of this pass, I'm not sure these need special handling. The pass will treat them as regular instructions, just with a large amount of registers used.

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 1 above)
Comment (ScottEgerton) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:90]: I didn't realise DenseSet wasn't ideal, I chose it as the elements need to be unique. Is there a better alternative that you would suggest?

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp (hunk 2 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:106]: I guess you should disable all optimization across a call, like you do across exec mask changes, but that can be a separate patch.

**Comment by jayfoad** [comment]
The test changes LGTM. I think this is ready to go if Matt is happy with DenseSet or a replacement (maybe just a SmallVector with a manual is_contained check before appending to it?).

**Review by jayfoad — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp [hunk 5]
```diff
@@ -83,37 +83,42 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
...
+          for (const MCRegUnit Unit : RegUnits)
             RegisterUseCount.erase(Unit);
         }
+
+        for (const auto &Operand : MI.all_uses()) {
+          const auto Reg = Operand.getReg();
+
+          // Count the number of times each register is read.
+          for (const MCRegUnit Unit : TRI->regunits(Reg)) {
+            if (!llvm::is_contained(RegistersUsed, Unit))
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:110]: Don't need the `llvm::`

**Pull request merged**